torch/optim/sgd.py

```diff
@@ -123,37 +128,174 @@ def step(self, closure=None):
             params_with_grad = []
             d_p_list = []
             momentum_buffer_list = []
-            weight_decay = group['weight_decay']
-            momentum = group['momentum']
-            dampening = group['dampening']
-            nesterov = group['nesterov']
-            maximize = group['maximize']
-            lr = group['lr']
+            has_sparse_grad = False
 
             for p in group['params']:
                 if p.grad is not None:
                     params_with_grad.append(p)
                     d_p_list.append(p.grad)
+                    if p.grad.is_sparse:
+                        has_sparse_grad = True
 
                     state = self.state[p]
                     if 'momentum_buffer' not in state:
                         momentum_buffer_list.append(None)
                     else:
                         momentum_buffer_list.append(state['momentum_buffer'])
 
-            F.sgd(params_with_grad,
-                  d_p_list,
-                  momentum_buffer_list,
-                  weight_decay=weight_decay,
-                  momentum=momentum,
-                  lr=lr,
-                  dampening=dampening,
-                  nesterov=nesterov,
-                  maximize=maximize,)
+            sgd(params_with_grad,
+                d_p_list,
+                momentum_buffer_list,
+                weight_decay=group['weight_decay'],
+                momentum=group['momentum'],
+                lr=group['lr'],
+                dampening=group['dampening'],
+                nesterov=group['nesterov'],
+                maximize=group['maximize'],
+                has_sparse_grad=has_sparse_grad,
+                foreach=group['foreach'])
 
             # update momentum_buffers in state
```

Comment:
This is necessary here because we pass None when initializing momentum_buffer, so the update in the functional does not alias. I didn't fold this into the functional and left it like this.

Is there a good way of passing a tensor of the shape of grad that signifies `None`? (would something zeros_like(p.grad) and then a zero check in the functional potentially cause issues?)

Reply:
I think we can significantly simplify all this by doing the initialization here (by initializing with `dampening / momentum * grad` and removing all the special case in the functional variant). But yes that can be left as future work for now.